libcxx/test/std/algorithms/alg.modifying.operations/alg.swap/swap_ranges.pass.cpp

```diff
@@ -14,162 +14,133 @@
     }
+  };
+};
+#endif
+
+template <template <class> class Iter1, template <class> class Iter2>
+TEST_CONSTEXPR_CXX20 bool test_simple_cases() {
+  {
+    int a[2] = {1, 2};
+    int b[2] = {4, 5};
+    std::swap_ranges(Iter1<int*>(a), Iter1<int*>(a + 2), Iter2<int*>(b));
+    assert(a[0] == 4 && a[1] == 5);
+    assert(b[0] == 1 && b[1] == 2);
+  }
+  {
+    std::array<int, 3> a = {1, 2, 3}, a0 = a;
+    std::array<int, 3> b = {4, 5, 6}, b0 = b;
+    using It1 = Iter1<std::array<int, 3>::iterator>;
+    using It2 = Iter2<std::array<int, 3>::iterator>;
+    std::swap_ranges(It1(a.begin()), It1(a.end()), It2(b.begin()));
+    assert(a == b0);
+    assert(b == a0);
+  }
+  {
+    std::array<std::array<int, 2>, 2> a = {{{0, 1}, {2, 3}}}, a0 = a;
+    std::array<std::array<int, 2>, 2> b = {{{9, 8}, {7, 6}}}, b0 = b;
+    using It1 = Iter1<std::array<std::array<int, 2>, 2>::iterator>;
+    using It2 = Iter2<std::array<std::array<int, 2>, 2>::iterator>;
+    std::swap_ranges(It1(a.begin()), It1(a.end()), It2(b.begin()));
+    assert(a == b0);
+    assert(b == a0);
+  }
+  {
+    std::array<std::array<int, 3>, 3> a = {{{0, 1, 2}, {3, 4, 5}, {6, 7, 8}}}, a0 = a;
+    std::array<std::array<int, 3>, 3> b = {{{9, 8, 7}, {6, 5, 4}, {3, 2, 1}}}, b0 = b;
+    using It1 = Iter1<std::array<std::array<int, 3>, 3>::iterator>;
+    using It2 = Iter2<std::array<std::array<int, 3>, 3>::iterator>;
+    std::swap_ranges(It1(a.begin()), It1(a.end()), It2(b.begin()));
+    assert(a == b0);
+    assert(b == a0);
+  }
+
+  return true;
 }
 
-#if TEST_STD_VER > 17
-constexpr bool test_swap_constexpr()
-{
-    int i[3] = {1, 2, 3};
-    int j[3] = {4, 5, 6};
-    std::swap_ranges(i, i+3, j);
-    return i[0] == 4 &&
-           i[1] == 5 &&
-           i[2] == 6 &&
-           j[0] == 1 &&
-           j[1] == 2 &&
-           j[2] == 3;
+TEST_CONSTEXPR_CXX20 bool test() {
+  test_simple_cases<forward_iterator, forward_iterator>();
+  test_simple_cases<forward_iterator, bidirectional_iterator>();
+  test_simple_cases<forward_iterator, random_access_iterator>();
+  test_simple_cases<bidirectional_iterator, forward_iterator>();
+  test_simple_cases<bidirectional_iterator, bidirectional_iterator>();
+  test_simple_cases<bidirectional_iterator, random_access_iterator>();
+  test_simple_cases<random_access_iterator, random_access_iterator>();
+#if TEST_STD_VER >= 20
+  test_simple_cases<std::type_identity_t, std::type_identity_t>();
+#endif
+
+  types::for_each(types::forward_iterator_list<int*>(), TestPtr());
+
+#if TEST_STD_VER >= 11 && TEST_STD_VER <= 17
+  types::for_each(types::forward_iterator_list<std::unique_ptr<int>*>(), TestUniquePtr());
+#elif TEST_STD_VER == 20
+  if (!std::is_constant_evaluated())
+    types::for_each(types::forward_iterator_list<std::unique_ptr<int>*>(), TestUniquePtr());
+#elif TEST_STD_VER >= 23
+  types::for_each(types::forward_iterator_list<std::unique_ptr<int>*>(), TestUniquePtr());
+#endif
```

Comment:
```suggestion
if (TEST_STD_AT_LEAST_23_OR_RUNTIME_EVALUATED)
  types::for_each(types::forward_iterator_list<std::unique_ptr<int>*>(), TestUniquePtr());
```

This is defined in `test_macros.h`. I'm not a huge fan of it, but it does simplify this code a lot. The alternative would be

```
if (!TEST_IS_CONSTANT_EVALUATED || TEST_STD_VER >= 23)
  types::for_each(types::forward_iterator_list<std::unique_ptr<int>*>(), TestUniquePtr());
```

This may trigger tautological comparison warnings though.

Reply:
Thanks! This works really well!